openwisp_firmware_upgrader/tests/test_models.py

```diff
@@ -124,6 +146,61 @@ def test_device_fw_no_connection(self):
         else:
             self.fail('ValidationError not raised')
 
+    def test_device_fw_not_created_on_device_connection_save(self):
+        os_identifier = 'Linux'
+        org = self._get_org()
+        category = self._get_category(organization=org)
+        build1 = self._create_build(
+            category=category, version='0.1', os_identifier=os_identifier
+        )
+        image1a = self._create_firmware_image(
+            build=build1, type=REVERSE_FIRMWARE_IMAGE_MAP['YunCore XD3200']
+        )
+
+        with self.subTest('Device doesn\'t define os'):
+            d1 = self._create_device_with_connection(model=image1a.boards[0])
+            self.assertEqual(DeviceConnection.objects.count(), 1)
+            self.assertEqual(Device.objects.count(), 1)
+            self.assertEqual(DeviceFirmware.objects.count(), 0)
+            d1.delete()
+            Credentials.objects.all().delete()
+
+        with self.subTest('Device doesn\'t define model'):
+            d1 = self._create_device_with_connection(os=os_identifier)
+            self.assertEqual(DeviceConnection.objects.count(), 1)
+            self.assertEqual(Device.objects.count(), 1)
+            self.assertEqual(DeviceFirmware.objects.count(), 0)
+            d1.delete()
+            Credentials.objects.all().delete()
+
+        build1.os_identifier = None
+        build1.save()
+
+        with self.subTest('Build doesn\'t define os_identifier'):
+            d1 = self._create_device_with_connection(
+                model=image1a.boards[0], os=os_identifier
+            )
+            self.assertEqual(DeviceConnection.objects.count(), 1)
+            self.assertEqual(Device.objects.count(), 1)
+            self.assertEqual(DeviceFirmware.objects.count(), 0)
+
+    def test_device_fw_created_on_device_connection_save(self):
+        self.assertEqual(DeviceFirmware.objects.count(), 0)
+        self.assertEqual(Device.objects.count(), 0)
+        os_identifier = 'Linux'
```

Comment:
let's use a more realistic os identifier in our tests, just to make it more familiar to future devs, eg: `OpenWrt 19.07-SNAPSHOT r11061-6ffd4d8a4d`.